### PR TITLE
Fix two local chat-template fragility bugs: leaked [TOOL_CALLS] names and rejected mixed tool_result+text turns

### PR DESCRIPTION
**Pull request by jankeydadondc-byte**
## Summary

Two distinct fragility shapes observed running local models (LM Studio,
devstral-small-2-2512) through the native Anthropic Messages transport.
Both trace back to the same underlying cause: local models render
conversations through their own Jinja chat template (Mistral-family
here), which is stricter and more fragile than the real Anthropic API's
actual accepted wire format.

**1. Tool-name corruption.** A `tool_use` `content_block_start`'s `name`
occasionally arrives as `[TOOL_CALLS]Read` instead of `Read` — Mistral's
chat-template control token for signaling a tool call leaks into the
parsed name field. Intermittent, model-specific. fcc relayed the
corrupted name straight through to the client, which then tries to
invoke a tool literally named `[TOOL_CALLS]Read`.

**2. Rejected mixed `tool_result`+`text` user turns.** Claude Code
routinely appends reminder text blocks (`"CRITICAL: respond with text
only"`, `<system-reminder>` content, etc.) into the *same* user turn as a
`tool_result`. The real Anthropic API renders that fine. Devstral's
Mistral template (as implemented in llama.cpp/LM Studio) refuses to
render any user turn mixing a `tool_result` with a non-`tool_result`
block:

> Anthropic streaming error: Error rendering prompt with jinja template:
> "After the optional system message, conversation roles must alternate
> user and assistant roles except for tool calls and results."

Since every tool call in a coding agent loop comes back as
`tool_result` + an injected reminder block, this fires almost immediately
and repeatedly in real agent use — and, via fcc's retry-then-timeout
handling on the failure, can present to the user as an unrelated
client-side "operation timed out" symptom rather than what it actually
is.

(Loosely related in spirit, though a different provider/transport: #326
describes OpenRouter→non-Anthropic-backend message shapes that also
break on "role alternation"/"tool messages following tool_calls" —
consistent with this being a broader pattern across non-native-Anthropic
backends, not something specific to one model or provider.)

## Repro

Isolated both with minimal-pair curls straight at LM Studio (bypassing
fcc entirely, to rule out anything else in the pipeline):

- A user turn with a `tool_result` **alone** → renders OK.
- The same turn plus a trailing `text` block → the jinja template error
  above.
- Splitting into two consecutive messages (tool-role then user-role)
  **also** fails — that breaks role alternation a different way.
- The only shape every template renders cleanly: a user turn containing
  **only** `tool_result` block(s).

```bash
# Against LM Studio directly (adjust model to whatever you have loaded):
curl -s http://localhost:1234/v1/messages \
  -H "content-type: application/json" -H "anthropic-version: 2023-06-01" \
  -d '{
    "model": "your-local-model",
    "max_tokens": 100,
    "messages": [
      {"role": "user", "content": "list files"},
      {"role": "assistant", "content": [{"type":"tool_use","id":"t1","name":"ls","input":{}}]},
      {"role": "user", "content": [
        {"type": "tool_result", "tool_use_id": "t1", "content": "a.txt\nb.txt"},
        {"type": "text", "text": "CRITICAL: respond with text only"}
      ]}
    ]
  }'
# -> Anthropic streaming error: Error rendering prompt with jinja template:
#    "After the optional system message, conversation roles must alternate..."
```

## Fix

Both are narrow, targeted sanitizers at the native Anthropic
request/response boundary — not changes to the real Anthropic API
contract, and safe to apply universally since the transformed shape is
semantically identical to the original for any client that reads it:

1. `core/anthropic/native_sse_block_policy.py`: new `sanitize_tool_name()`
   strips a leading run of `[TOOL_CALLS]`/`[/TOOL_CALLS]` control tokens
   from `tool_use` `content_block_start` names, applied at the earliest
   transform point (`transform_native_sse_block_event`).
2. `core/anthropic/native_messages_request.py`: new
   `sanitize_tool_result_user_messages()` folds any non-`tool_result`
   blocks in a user turn into the **last** `tool_result`'s content
   (string content gets the text appended with a blank-line join; list
   content gets the blocks appended), leaving a `tool_result`-only
   turn — the model reads the identical text either way. Wired into
   `build_base_native_anthropic_request_body()` for all native-transport
   providers.

## Testing

- `uv run ruff format` / `ruff check --fix` / `ty check`: all clean on
  every changed file.
- Added cases to `tests/core/anthropic/test_native_sse_block_policy.py`
  for `sanitize_tool_name()` directly and through the real
  `transform_native_sse_block_event()` SSE-transform path with the exact
  recorded corruption (`[TOOL_CALLS]Read` → `Read`).
- Added `tests/core/anthropic/test_native_messages_request_tool_result_folding.py`
  covering: `tool_result`-only passthrough, string-content folding,
  list-content folding, multiple `tool_result`s in one turn (extras fold
  into the last one only), plain text/assistant messages as no-ops,
  non-list input passthrough, and a non-text extra (e.g. an image)
  preserved as its own block rather than merged into text.
- Manually verified every new test's assertions by direct execution
  outside pytest — see the note in my other two PRs from this same
  session (#976, #977) on a pre-existing, environment-specific pytest
  crash on Windows, confirmed unrelated to any of these changes.

## Scope

Third in a small series from the same debugging session (after #976,
#977) — bundling these two together since both are the same class of
fix ("local chat-template fragility").

Version bumped `2.4.4` → `2.4.5` per `AGENTS.md`'s versioning rule (bug
fix, no breaking change). Note: #976 and #977 also bump `2.4.4` →
`2.4.5` independently — whichever of these three lands last will hit a
trivial version-string conflict in `pyproject.toml`/`uv.lock`; happy to
rebase immediately if that happens.

## Versioning note

The semver bump is intentionally omitted from this PR (same as #976). `main`
merges several version-bumping PRs a day, so any version this branch pins goes
stale — and re-conflicts the PR — within hours. This rebase also folded in
upstream's move of cloud providers to the OpenAI-chat transport (the OpenRouter
native helpers this branch's file used to sit next to are gone); the code diff
here is now just the two local-chat-template fixes plus their tests. Happy to
push a bump at merge time if you ping me, or bump on merge — your call.

<h3>Greptile Summary</h3>

This PR adds sanitizers for local Anthropic-compatible chat-template edge cases. The main changes are:

- Mixed user turns with `tool_result` blocks are folded into tool-result-only content before native request forwarding.
- Tool-result folding preserves leading, trailing, and interleaved block order.
- Streamed `tool_use` block names have leaked `[TOOL_CALLS]` control tokens stripped before reaching the client.
- Unit tests cover request folding and SSE tool-name sanitization paths.

<h3>Confidence Score: 4/5</h3>

The sanitizer changes are narrow and covered by focused tests.

The request folding preserves block order and the SSE tool-name cleanup is applied at the native boundary.

`pyproject.toml` and `uv.lock` need the required version bump for the production changes.

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- We ran the anthropic sanitizers test suite with pytest and it completed successfully with 17 tests passed.
- Lint and type-check steps completed with success (ruff and ty exit codes were 0).
- We reviewed the sanitizer probe output to confirm the direct before/after traces for Read tool calls and mixed tool\_result plus text scenarios.

<a href="https://app.greptile.com/trex/runs/13879201/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/free_claude_code/core/anthropic/native_messages_request.py | Adds native request sanitization that folds mixed user `tool_result` turns into tool-result-only content while preserving order. |
| src/free_claude_code/core/anthropic/native_sse_block_policy.py | Adds `tool_use` SSE start-block name sanitization to remove leaked local chat-template control tokens. |
| tests/core/anthropic/test_native_messages_request_tool_result_folding.py | Adds unit tests for passthrough, string/list folding, multiple tool results, and order-preserving leading/interleaved extras. |
| tests/core/anthropic/test_native_sse_block_policy.py | Extends SSE policy tests for direct tool-name sanitization and the real `content_block_start` transform path. |

<h3>Sequence Diagram</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
participant Client as Claude Code client
participant FCC as free-claude-code native Anthropic transport
participant Req as native_messages_request
participant Model as Local Anthropic-compatible model
participant SSE as native_sse_block_policy

Client->>FCC: Messages request with tool_result + reminder text
FCC->>Req: build_base_native_anthropic_request_body()
Req->>Req: sanitize_tool_result_user_messages()
Req-->>FCC: user turn contains only tool_result blocks
FCC->>Model: Native Anthropic request
Model-->>FCC: SSE content_block_start(tool_use name may include [TOOL_CALLS])
FCC->>SSE: transform_native_sse_block_event()
SSE->>SSE: sanitize_tool_name()
SSE-->>Client: content_block_start with clean tool name
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
participant Client as Claude Code client
participant FCC as free-claude-code native Anthropic transport
participant Req as native_messages_request
participant Model as Local Anthropic-compatible model
participant SSE as native_sse_block_policy

Client->>FCC: Messages request with tool_result + reminder text
FCC->>Req: build_base_native_anthropic_request_body()
Req->>Req: sanitize_tool_result_user_messages()
Req-->>FCC: user turn contains only tool_result blocks
FCC->>Model: Native Anthropic request
Model-->>FCC: SSE content_block_start(tool_use name may include [TOOL_CALLS])
FCC->>SSE: transform_native_sse_block_event()
SSE->>SSE: sanitize_tool_name()
SSE-->>Client: content_block_start with clean tool name
```

</a>

<h3>Comments Outside Diff (1)</h3>

1. `core/anthropic/native_messages_request.py`, line 365-368 ([link](https://github.com/alishahryar1/free-claude-code/blob/f0b01a3fca0560d27231f28a7bc25b848e5066d5/core/anthropic/native_messages_request.py#L365-L368)) 

   <a href="#">P1</a> **Apply folding consistently**
   `build_openrouter_native_request_body()` still sends `body["messages"]` through only the thinking sanitizer, while `request_policy.py` routes `extra_body == "openrouter"` providers exclusively to this builder. A mixed `tool_result` + `text` turn on that native OpenRouter path bypasses the new folding sanitizer and reaches the upstream in the rejected shape this PR is fixing.

   <details><summary><strong>Artifacts</strong></summary><br />

   **[Repro: focused Python harness for OpenRouter tool\_result folding](https://app.greptile.com/trex/artifacts/a5d50e61-18cb-487b-822a-d4ff367acb20)**

   - Contains supporting evidence from the run (text/x-python; charset=utf-8).

   **[Repro: execution output showing OpenRouter mixed blocks remain unfurled](https://app.greptile.com/trex/artifacts/4360e51a-7fb0-4ace-8952-4f6f4f06ff95)**

   - Keeps the command output available without making the summary code-heavy.

   <a href="https://app.greptile.com/trex/runs/13303900/artifacts?artifact=a5d50e61-18cb-487b-822a-d4ff367acb20"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewArtifacts.svg?v=4">View artifacts</picture></a>
   </details>

   <sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>

<sub>Reviews (4): Last reviewed commit: ["Fix two local chat-template fragility bu..."](https://github.com/alishahryar1/free-claude-code/commit/83695afe840d804582bb44d3c7dcd7c635b2c34d) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41864041)</sub>

### CODE REVIEW

File: core/anthropic/native_messages_request.py [hunk 1]
```diff
@@ -185,6 +185,84 @@ def sanitize_native_messages_thinking_policy(
...
+            block
+            for block in content
+            if not (isinstance(block, dict) and block.get("type") == "tool_result")
+        ]
+        if not tool_results or not extras:
+            sanitized_messages.append(message)
+            continue
+
+        new_results = [dict(block) for block in tool_results]
+        _fold_blocks_into_tool_result(new_results[-1], extras)
```
[the remark this answers is not in the document]
Comment (jankeydadondc-byte) [reply core/anthropic/native_messages_request.py]: Fixed (commit f0b01a3): the fold is now order-preserving. Instead of collecting all non-tool_result blocks and appending them to the last tool_result, it walks the blocks in order and folds each into an adjacent tool_result by position — blocks before a tool_result go ahead of its content, blocks after the last one go after. So a leading text block now stays before the tool output. Added regression tests for leading text, interleaved text across multiple tool_results, and a leading image; all prior tests still pass (their extras were all trailing).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by Ashok98765vvs** [comment]
## Review: Fix for Local Chat-Template Fragility Bugs

**Confidence Score: 4/5** — Mostly safe and well-targeted, with one known gap remaining.

### What This PR Fixes
This PR addresses two real fragility issues observed when running local models (LM Studio / devstral-small-2-2512) through the native Anthropic Messages transport:

1. **Tool-name corruption** — `[TOOL_CALLS]Read` leaking into parsed tool names due to Mistral control tokens bleeding into the `name` field.
2. **Rejected mixed `tool_result + text` user turns** — Mistral Jinja templates are stricter than the real Anthropic API and reject user messages that mix `tool_result` blocks with plain text.

### What Looks Good
- `sanitize_tool_name()` is narrow, correct, and well-tested — strips `[TOOL_CALLS]` / `[/TOOL_CALLS]` prefix cleanly.
- `sanitize_tool_result_user_messages()` folding logic is safe — it reshapes the wire format without losing any content the model needs to see.
- Regression tests cover the key cases: passthrough, string folding, list folding, multi-tool-result turns, and non-text block preservation (images).
- All required CI checks (pytest, ruff-check, ruff-format, type ignore suppressions) are passing.

### Remaining Gap
- `build_openrouter_native_request_body()` currently bypasses the new `tool_result` folding sanitizer — mixed `tool_result + text` shapes can still escape on the OpenRouter native path. This should be patched in a follow-up before this is considered fully resolved across all non-Anthropic backends.

### Blockers Before Merge
- Branch conflict must be resolved (`Resolve conflicts` button visible).
- `Ban suppressions and legacy annotations` check is still failing — needs attention.

Overall this is a solid, focused fix. Once the conflict is resolved and the OpenRouter path is also covered, this is safe to merge. Thanks for the detailed summary and tests!

File: core/anthropic/native_messages_request.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jankeydadondc-byte) [reply core/anthropic/native_messages_request.py]: Fixed (commit f0b01a3): `_fold_blocks_into_tool_result` now takes explicit `leading`/`trailing` lists and rebuilds content in order instead of appending everything to the last tool_result. A leading text block now stays before the tool output; a trailing reminder stays after (unchanged). Verified directly against your repro — `[text, tool_result]` now folds to `"read this first\n\ntool output here"` in that order. Added regression tests for leading text, interleaved text across multiple tool_results, and a leading image.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by jankeydadondc-byte** [comment]
@Alishahryar1 — rebased this cleanly onto the new `src/free_claude_code/` layout (#1029), so it's back to green on all required checks (Ban suppressions, ruff-format, ruff-check, ty, pytest) and is currently **mergeable**. The change is minimal: two self-contained local-chat-template fixes plus their tests, no version files touched.

One open item that's genuinely your call: the versioning policy asks for a `pyproject.toml` bump in the same commit, but pinning a version here just re-conflicts the PR each time `main` bumps (it's happened several times already). I've deliberately left the bump out to keep the PR mergeable — I'm happy to push the version bump the moment before you merge, or you're welcome to bump on merge, whichever you prefer.

Could you give this (and its sibling #976) a review when you have a moment? Both are green and ready. Thanks!

_🤖 Posted via automation on behalf of the PR author._

**Comment by jankeydadondc-byte** [comment]
Withdrawing this PR per our own review-cadence policy — no maintainer bandwidth issue implied. We'll keep maintaining the fix in our local patch set; happy to reopen or rebase if there's interest later. Thanks for maintaining the project!
